### PR TITLE
feat(FEC-12865) | Exposed ability to pass post params on DRM license

### PR DESCRIPTION
**Pull request by GouravSna**
### Description of the Changes

This feature gives ability to FE apps to pass `POST` parameters on the DRM license request. It can be done using `PKRequestParams.Adapter`'s `buildDrmPostParams` implementation.

> `Uri uri = requestParams.url.buildUpon().build(); ` URI is license URL which is important to pass while using `Adapter`.
    FE can pass the `QUERY` params on DRM license as well. `requestParams.url.buildUpon().appendQueryParameter("CustomParam", "MyQueryParam").build();`
   
>   - In case if the license URL is not passed then FE will get a Player Error event,
     ` errorType = DRM_ERROR errorMessage = java.lang.IllegalStateException: No license URL-ERROR_CODE_DRM_LICENSE_ACQUISITION_FAILED`

> - In case if the `POST` params are null then FE will get Player Error event,
    `errorType = DRM_ERROR errorMessage = com.kaltura.android.exoplayer2.upstream.HttpDataSource$InvalidResponseCodeException: Response code: 400-ERROR_CODE_DRM_LICENSE_ACQUISITION_FAILED`

#### Steps to pass `POST` parms

1. Create a class implementing `PKRequestParams.Adapter`.
2. Implement `PKRequestParams adapt(PKRequestParams requestParams)` method and modify the license URL if required for `QUERY` params otherwise send as it is.
3. Create `PKRequestParams` object by passing the license URI and headers if required.
4. Call `setPostBody` and pass `POST` body `Map` object.
5. Implement `JSONObject buildDrmPostParams(@NonNull byte[] data)` method and create a `JSONObject` for the `POST` params. FE apps can use `byte[] data` for further purposes. Return the `JSONObject`.
6. Now create an instance of this class and pass it to the `PlayerSettings`.
    Example:
    ```java
    DRMAdapter licenseRequestAdapter = new DRMAdapter();
    player.getSettings().setLicenseRequestAdapter(licenseRequestAdapter);
    ```

Example:

```java
static class DRMAdapter implements PKRequestParams.Adapter {
        String TOKEN = "YOUR_CUSTOM_TOKEN_FOR_DRM_LICENSE_POST_PARAM";

        public static String customData;
        @Override
        public PKRequestParams adapt(PKRequestParams requestParams) {
            Map<String, String> headers = new HashMap<>();
            Map<String, String> postBody = new HashMap<>();
            postBody.put("token", TOKEN);
            Uri uri = requestParams.url.buildUpon().appendQueryParameter("CustomParam", "MyQueryParam").build();
            PKRequestParams params = new PKRequestParams(uri, headers);
            params.setPostBody(postBody);
            return params;
        }

        @Override
        public void updateParams(Player player) {
            // TODO
        }

        @Override
        public String getApplicationName() {
             // TODO
        }

        @Override
        public JSONObject buildDrmPostParams(@NonNull byte[] data) {
            try {
                return // Create a JSONObject for the post body for DRM license request.;
            } catch (JSONException e) {
                return null;
            }
        }
    }

DRMAdapter licenseRequestAdapter = new DRMAdapter();
player.getSettings().setLicenseRequestAdapter(licenseRequestAdapter);

```

### CODE REVIEW

File: playkit/src/main/java/com/kaltura/playkit/drm/DrmCallback.java [hunk 1]
```diff
@@ -1,43 +1,88 @@
...
     }
 
-    public DrmCallback(HttpDataSource.Factory dataSourceFactory, PKRequestParams.Adapter adapter) {
-        this.dataSourceFactory = dataSourceFactory;
-        this.adapter = adapter;
+    @NonNull
+    @Override
+    public byte[] executeKeyRequest(@NonNull UUID uuid, @NonNull ExoMediaDrm.KeyRequest request) throws MediaDrmCallbackException {
+        if (adapter != null && !postBodyMap.isEmpty()) {
+            if (TextUtils.isEmpty(licenseUrl)) {
```
Comment (GouravSna) [inline playkit/src/main/java/com/kaltura/playkit/drm/DrmCallback.java:56]: Check if license is being `null`.

File: playkit/src/main/java/com/kaltura/playkit/drm/DrmCallback.java [hunk 2]
```diff
@@ -1,43 +1,88 @@
...
+                            : (C.CLEARKEY_UUID.equals(uuid) ? "application/json" : "application/octet-stream");
+            requestProperties.put("Content-Type", contentType);
+            if (MediaSupport.PLAYREADY_UUID.equals(uuid)) {
+                requestProperties.put(
+                        "SOAPAction", "http://schemas.microsoft.com/DRM/2007/03/protocols/AcquireLicense");
+            }
+            if (!headers.isEmpty()) {
+                requestProperties.putAll(headers);
+            }
+            JSONObject postBodyJsonObject = adapter.updateDRMData(request.getData());
```
Comment (GouravSna) [inline playkit/src/main/java/com/kaltura/playkit/drm/DrmCallback.java]: Check by passing null to JSONObject

File: playkit/src/main/java/com/kaltura/playkit/drm/DrmCallback.java [hunk 3]
```diff
@@ -1,43 +1,88 @@
...
+            return executePost(dataSourceFactory, licenseUrl, postBodyJsonObject, requestProperties);
+        } else {
+            return callback.executeKeyRequest(uuid, request);
```
Comment (GouravSna) [inline playkit/src/main/java/com/kaltura/playkit/drm/DrmCallback.java:81]: Check with kaltura drm media.

File: playkit/src/main/java/com/kaltura/playkit/drm/DrmCallback.java (hunk 2 above)
Comment (GouravSna) [reply playkit/src/main/java/com/kaltura/playkit/drm/DrmCallback.java]: Done. Update the PR description with the error sent to FE in this case.

File: playkit/src/main/java/com/kaltura/playkit/drm/DrmCallback.java (hunk 1 above)
Comment (GouravSna) [reply playkit/src/main/java/com/kaltura/playkit/drm/DrmCallback.java:56]: Done. Update the PR description with the error sent to FE in this case.

File: playkit/src/main/java/com/kaltura/playkit/drm/DrmCallback.java [hunk 4]
```diff
@@ -1,43 +1,88 @@
...
 import java.util.UUID;
 
 public class DrmCallback implements MediaDrmCallback {
     private static final PKLog log = PKLog.get("DrmCallback");
-
     private final HttpDataSource.Factory dataSourceFactory;
     private final PKRequestParams.Adapter adapter;
     private HttpMediaDrmCallback callback;
+    private final Map<String, String> postBodyMap = new HashMap<>();
+    private String licenseUrl;
```
Comment (giladna) [inline playkit/src/main/java/com/kaltura/playkit/drm/DrmCallback.java:37]: can we flip the lines?
   private String licenseUrl;
   private final Map<String, String> headers = new HashMap<>();;
   private final Map<String, String> postBodyMap = new HashMap<>();

Comment (giladna) [reply playkit/src/main/java/com/kaltura/playkit/drm/DrmCallback.java:37]: @GouravSna

File: playkit/src/main/java/com/kaltura/playkit/drm/DrmCallback.java [hunk 5]
```diff
@@ -1,43 +1,88 @@
...
+            JSONObject postBodyJsonObject = adapter.buildDrmPostParams(request.getData());
+            return executePost(dataSourceFactory, licenseUrl, postBodyJsonObject, requestProperties);
+        } else {
+            return callback.executeKeyRequest(uuid, request);
+        }
     }
 
     void setLicenseUrl(String licenseUrl) {
-
         if (licenseUrl == null) {
```
Comment (giladna) [inline playkit/src/main/java/com/kaltura/playkit/drm/DrmCallback.java]: licenseUrl is confusing with this.licenseUrl lets keep bit different names

**Review by giladna — APPROVED** [review]

**Pull request merged**